### PR TITLE
fixed gitlab integration

### PR DESCRIPTION
**Pull request by Zveroboi123**
1. gitlab groups were fixed 
2. gitlab authentication fixed - gitlab expects only auth token. no username required

### CODE REVIEW

File: source/gitlab/gitlab.go [hunk 1]
```diff
@@ -8,11 +8,10 @@ import (
 	"net/http"
 	nurl "net/url"
 	"os"
+	"path"
 	"strconv"
 	"strings"
-)
 
-import (
```
Comment (dhui) [inline source/gitlab/gitlab.go]: Leave the import grouping as is

**Comment by Zveroboi123** [comment]
Empty or dummy username would work just fine, of course, but might confuse users. The main fix was around the GitLab groups. Would you like to leave the previous format - "username@password" in place?

File: source/gitlab/gitlab.go (hunk 1 above)
Comment (Zveroboi123) [reply source/gitlab/gitlab.go]: Sure

**Comment by dhui** [comment]
> Would you like to leave the previous format - "username:password" in place?

Yes, leave the original URI format in place.

> The main fix was around the GitLab groups.

Your change is backwards incompatible. Have you tried URL encoding the `/` for your group?

**Comment by Zveroboi123** [comment]
I tried encoding it, but it gets automatically decoded and treated as regular slash. using backslash didn't work either.
I'd like to propose a new backward compatible format: gitlab://username:token@<domain>/<group>\\<project>/<sub-folder>
so that it would work either way. We could think of another group-project delimiter, if you like. 
What do you think?

**Comment by Zveroboi123** [comment]
Should I take care of the coverage degradation?
It's just 0.08% and is probably due to adding a single line of code.

**Comment by dhui** [comment]
> I tried encoding it, but it gets automatically decoded and treated as regular slash. using backslash didn't work either.
I'd like to propose a new backward compatible format: gitlab://username:token@<domain>/<group>\\<project>/<sub-folder>
so that it would work either way. We could think of another group-project delimiter, if you like.
What do you think?

What's the valid character set for a username or group? I couldn't find any documentation about it.
This was the closest I could find: https://docs.gitlab.com/ee/user/reserved_names.html
If we're going to pick a character in the valid character set, we'll need to be able to escape it...
e.g. `\/` => `/` and `\\` => `\`

> Should I take care of the coverage degradation?

Thanks for being aware of this. I wouldn't worry about the decrease in code coverage at this point.

**Comment by Zveroboi123** [comment]
Got your point. I added escaping support and verified it is backward compatible.
Following formats will work fine:
1. gitlab://aaa:kZX7yRFHa8AMyKjYtTc3@gitlab.contoso.com/my\\\\group\\myproject/myprojectpath#master
2. gitlab://aaa:kZX7yRFHa8AMyKjYtTc3@gitlab.contoso.com/mygroup\\myproject/myprojectpath#master
3. 
gitlab://aaa:kZX7yRFHa8AMyKjYtTc3@gitlab.contoso.com/myproject/myprojectpath#master

**Comment by dhui** [comment]
Actually, instead adding our own escaping to the DSN, use [url.URL.RawPath](http://localhost:8088/pkg/net/url/#URL) instead of `url.URL.Path`. That way, URL-encoded strings in the path should work.

**Comment by Zveroboi123** [comment]
Hi, I don't fully understand you. I'm not inventing my own escaping here.
I'm using regex to split group + projectId from the rest of the path. All
unescaping is done automatically. Just run my code to see how it works.

**Comment by dhui** [comment]
> Hi, I don't fully understand you. I'm not inventing my own escaping here.

The current approach of using regexs is not very clean. My original though was that we'd be adding backslash escaping on top of url encoding which is also dirty...
I believe both of these dirty approaches can be avoided by using the escaped path.

> Please check out the following thread: https://stackoverflow.com/questions/39270633/why-does-url-parse-not-populate-url-rawpath#

Ahh, yeah, it looks like we should be using `EscapedPath()` instead.
https://play.golang.org/p/V838Xn_OZMa

so `pe := strings.Split(strings.Trim(u.Path, "/"), "/")` would become `pe := strings.Split(strings.Trim(u.EscapedPath(), "/"), "/")`

> current gitlab implementation is pretty much useless for 95% of us.

No idea about the data for this... e.g. number of users using groups/projects Support for gitlab sources was only merged ~2 days ago.
Obviously, the original author didn't have this problem.

> it would be super cool if you would approve it as it already is to unblock us all.

If you're blocked, use your fork.

Also, **whatever the final solution is, we should document it** so future users aren't confused.

**Comment by Zveroboi123** [comment]
Hi,
I personally don't like the approach to have a slash delimiter encoded
between group and the project Id. I think the group name itself should be
encoded if it contains a slash within.
Unfortunately, I don't have time to work on it. Will be using my fork.
Thank you and good luck

**Pull request closed without merging**